icu4c/source/stubdata/stubdata.cpp

```diff
@@ -19,26 +19,8 @@
 *   for running the data building tools.
 *
 */
-#include "unicode/utypes.h"
-#include "unicode/udata.h"
-#include "unicode/uversion.h"
 
-
-typedef struct {
-    uint16_t headerSize;
-    uint8_t magic1, magic2;
-    UDataInfo info;
-    char padding[8];
-    uint32_t count, reserved;
-    /*
-    const struct {
-    const char *const name; 
-    const void *const data;
-    } toc[1];
-    */
-   int   fakeNameAndData[4];       /* TODO:  Change this header type from */
-                                   /*        pointerTOC to OffsetTOC.     */
-} ICU_Data_Header;
+#include "stubdata.h"
```

Comment:
Why is it necessary to pull these definitions into a new header file that is only included from here?

Reply:
`U_ICUDATA_ENTRY_POINT` is already exported and referenced when building `libicuuc`. My understanding is that all exported symbols should have a matching declaration in a header file. This is specifically a requirement for TAPI, which is enabled in Apple's ICU fork.

Reply:
stubdata is an artificial construct. Its structure matches other definitions, and matches expectations by the runtime code, but the precise struct is unique. Moving it to a header file seems harmless, but also unnecessary, since no one else is including it.

I don't remember quite why this struct definition is unique. It's been a long time...

Reply:
> ... why this struct definition is unique...

FWIW, I found an older ticket that might give some insight: https://unicode-org.atlassian.net/browse/ICU-8043

From the ticket (emphasis added):

> U_ICUDATA_ENTRY_POINT is defined at 2 different places as 2 differents structure instance. It prevents a compiler (Visual C++) from inlining across different compilation units.

> The two mismatched struct types in question, one is a data header only (the declaration that the library code sees) and one is the same data header followed by some canned data.  **That canned data won't be present in all cases, so I don't want to make its declaration visible to the main library code.**

Reply:
Thanks for the background! If I understand correctly, this object is declared additionally https://github.com/unicode-org/icu/blob/85705f04e03f9cb41d4406bcd38c3e419eb7889d/icu4c/source/common/udata.cpp#L646 and also referenced in that file. 
I would naively suggest that this is where `stubdata.h` should be included. Attempting to do so reflects 
```
/icu/icuSources/common/udata.cpp:702:52: error: comparison of distinct pointer types ('const DataHeader *' and 'const ICU_Data_Header *')
                if(gCommonICUDataArray[i]->pHeader == &U_ICUDATA_ENTRY_POINT) {
````

However `ICU_Data_Header` ends up being the type of `U_ICUDATA_ENTRY_POINT` at runtime since it links against subdata's definition (at least on Darwin). It sounds like it's intentionally this way, but it seems a bit concerning that the compiler is tricked, and sees a different type for the object.

Reply:
> However `ICU_Data_Header` ends up being the type of `U_ICUDATA_ENTRY_POINT` at runtime since it links against subdata's definition (at least on Darwin). It sounds like it's intentionally this way, but it seems a bit concerning that the compiler is tricked, and sees a different type for the object.

It's a semi-arbitrary type for an engineered blob of bytes. The data format is variable-width to accommodate anything that we throw at it. When you build the ICU data into a shared library, then you link with that instead of with the stubdata. The contents of the shared library changes with the input data. The shared library itself can be created directly from a .o file (at least on Linux with ELF objects, and similarly on Windows), where there is no type info at all, just a linker symbol pointing to the bytes, or it can be created from a C or assembler source file.

https://unicode-org.github.io/icu/userguide/icu_data/#data-in-shared-librariesdlls-vs-dat-package-files

The compiler should not care about the exact C++ data type. If it does, it's best to silence that for the data entry point.

Reply:
I see, thanks for the explanation. Though in principle I believe its good to have declarations in headers reflect what is exposed in a binary's symbol table, I don't feel strongly about keeping this header if it will just become a maintenance burden without a clear benefit. Maybe when/if ICU adopts TAPI validation this could be revisited. What do you think is best for moving forward with this PR?

Reply:
I'll update with a comment to specify this situation. 

> If it's possible to add a line that silences the "declarations should be in a header file" warning for the struct in stubdata.cpp, then that would be best.

This is possible with TAPI, but the true power of tapi is using these header declarations to generate artifacts that can be directly consumed by our system linker (ld64). This allows for increased parallelization across shared object dependencies. Demoting this error into a warning means this symbol would no longer be reference-able when that artifact exists.

Reply:
Added in: https://github.com/unicode-org/icu/pull/2118/files#diff-dfab250a2804b93e2e4135dfc82d4df6f5f2d55e5490379d2af12cbe962602f4R11